src/main/java/com/metamx/common/CompressionUtils.java

```diff
@@ -0,0 +1,388 @@
+    try {
+      FileUtils.FileCopyResult copyResult = FileUtils.retryCopy(
+          byteSource,
+          tmpFile,
+          shouldRetry,
+          DEFAULT_RETRY_COUNT
+      );
+      return unzip(tmpFile, outDir);
+    }
+    finally {
+      if (!tmpFile.delete()) {
+        log.warn("Could not delete zip cache at [%s]", tmpFile.toString());
+      }
+    }
+  }
+
+  public static FileUtils.FileCopyResult unzip(
+      final ByteSource byteSource,
+      final File outDir,
+      boolean cacheLocally
+  ) throws IOException
+  {
+    return unzip(byteSource, outDir, Predicates.<Throwable>alwaysTrue(), cacheLocally);
+  }
+
+  public static FileUtils.FileCopyResult unzip(final File pulledFile, final File outDir) throws IOException
+  {
+    if (!(outDir.exists() && outDir.isDirectory())) {
+      throw new ISE("outDir[%s] must exist and be a directory", outDir);
+    }
+    log.info("Unzipping file[%s] to [%s]", pulledFile, outDir);
+    final FileUtils.FileCopyResult result = new FileUtils.FileCopyResult();
+    try (final ZipFile zipFile = new ZipFile(pulledFile)) {
+      final Enumeration<? extends ZipEntry> enumeration = zipFile.entries();
+      while (enumeration.hasMoreElements()) {
+        final ZipEntry entry = enumeration.nextElement();
+        result.addFiles(
+            FileUtils.retryCopy(
+                new ByteSource()
+                {
+                  @Override
+                  public InputStream openStream() throws IOException
+                  {
+                    return new BufferedInputStream(zipFile.getInputStream(entry));
+                  }
+                },
+                new File(outDir, entry.getName()),
+                Predicates.<Throwable>alwaysTrue(),
+                DEFAULT_RETRY_COUNT
+            ).getFiles()
+        );
+      }
+    }
+    return result;
+  }
+
+  // Use unzip(ByteStream, File) if possible
+  public static FileUtils.FileCopyResult unzip(InputStream in, File outDir) throws IOException
+  {
+    try (final ZipInputStream zipIn = new ZipInputStream(in)) {
+      final FileUtils.FileCopyResult result = new FileUtils.FileCopyResult();
+      ZipEntry entry;
+      while ((entry = zipIn.getNextEntry()) != null) {
+        final File file = new File(outDir, entry.getName());
+        Files.asByteSink(file).writeFrom(zipIn);
+        result.addFile(file);
+        zipIn.closeEntry();
+      }
+      return result;
+    }
+  }
+
+  /**
+   * gunzip the file to the output. If out is a directory, it will gunzip the pulled file using the name less `.gz`.
+   * If out is a not a directory, it will simply gunzip to that file
+   *
+   * @param pulledFile The source of the gz data
+   * @param out        Either a directory or a target file
+   *
+   * @return The result of the file copy
+   *
+   * @throws IOException
+   */
+  public static FileUtils.FileCopyResult gunzip(final File pulledFile, File out) throws IOException
+  {
+    if (out.isDirectory()) {
```

Comment:
Is there a reason that we need to switch behavior based on whether or not "out" is a directory? I think those kinds of magical APIs make code more confusing, because it's not always clear from the call site whether the second parameter is supposed to be a directory or file, and that makes it harder to understand.